comm/src/main/scala/coop/rchain/comm/rp/Connect.scala

```diff
@@ -127,20 +130,23 @@ object Connect {
 
   def findAndConnect[F[_]: Sync: Monad: Log: Time: Metrics: NodeDiscovery: ErrorHandler: ConnectionsCell: RPConfAsk](
       conn: (PeerNode, FiniteDuration) => F[Unit]
-  ): F[List[PeerNode]] =
-    for {
-      connections      <- ConnectionsCell[F].read
-      tout             <- RPConfAsk[F].reader(_.defaultTimeout)
-      peers            <- NodeDiscovery[F].peers.map(p => (p.toSet -- connections).toList)
-      responses        <- peers.traverse(p => ErrorHandler[F].attempt(conn(p, tout)))
-      peersAndResonses = peers.zip(responses)
-      _ <- peersAndResonses.traverse {
-            case (peer, Left(error)) =>
-              Log[F].debug(s"Failed to connect to ${peer.toAddress}. Reason: ${error.message}")
-            case (peer, Right(_)) =>
-              Log[F].info(s"Connected to ${peer.toAddress}.")
-          }
-    } yield peersAndResonses.filter(_._2.isRight).map(_._1)
+  ): F[List[PeerNode]] = {
+    def find(): F[List[PeerNode]] =
+      for {
+        connections      <- ConnectionsCell[F].read
+        tout             <- RPConfAsk[F].reader(_.defaultTimeout)
+        peers            <- NodeDiscovery[F].peers.map(p => (p.toSet -- connections).toList)
+        responses        <- peers.traverse(p => ErrorHandler[F].attempt(conn(p, tout)))
+        peersAndResonses = peers.zip(responses)
```

Comment:
So if attempting to connect is asynchronous how come you're waiting for a response here?

Also you have a typo in `peersAndResonses`

Reply:
@KentShikama Actually, this code is not new and not the significant change of the PR. ;-)
We are using `TransportLayer.send` (Handshake) to connect to other peers. The current implementation of `send` is not _fire-and-forget_. There is a confirmation of delivery. If the message can't be delivered to the remote peer then `TransportLayer.send` returns a `CommError`. But there is no guarantee that the delivered message has been or will get processed by the remote peer (perhaps besides that the message has been added to a message queue).

Reply:
But now I see that this will be a problem
`NodeDiscovery[F].peers.map(p => (p.toSet -- connections).toList)`

We should take only so many peers we have space left in the connection list. I'll fix that.

Reply:
bors r+